crates/ra_assists/src/add_missing_impl_members.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{Assist, AssistId, AssistCtx};
+
+use hir::Resolver;
+use hir::db::HirDatabase;
+use ra_syntax::{SmolStr, SyntaxKind, TextRange, TextUnit, TreeArc};
+use ra_syntax::ast::{self, AstNode, FnDef, ImplItem, ImplItemKind, NameOwner};
+use ra_db::FilePosition;
+use ra_fmt::{leading_indent, reindent};
+
+use itertools::Itertools;
+
+/// Given an `ast::ImplBlock`, resolves the target trait (the one being
+/// implemented) to a `ast::TraitDef`.
+pub(crate) fn resolve_target_trait_def(
+    db: &impl HirDatabase,
+    resolver: &Resolver,
+    impl_block: &ast::ImplBlock,
+) -> Option<TreeArc<ast::TraitDef>> {
+    let ast_path = impl_block.target_trait().map(AstNode::syntax).and_then(ast::PathType::cast)?;
+    let hir_path = ast_path.path().and_then(hir::Path::from_ast)?;
+
+    match resolver.resolve_path(db, &hir_path).take_types() {
+        Some(hir::Resolution::Def(hir::ModuleDef::Trait(def))) => Some(def.source(db).1),
+        _ => None,
+    }
+}
+
+pub(crate) fn build_func_body(def: &ast::FnDef) -> String {
+    let mut buf = String::new();
+
+    for child in def.syntax().children() {
+        if child.kind() == SyntaxKind::SEMI {
+            buf.push_str(" { unimplemented!() }")
+        } else {
+            child.text().push_to(&mut buf);
+        }
+    }
+
+    buf.trim_end().to_string()
+}
+
+pub(crate) fn add_missing_impl_members(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let node = ctx.covering_node();
+    let impl_node = node.ancestors().find_map(ast::ImplBlock::cast)?;
+    let impl_item_list = impl_node.item_list()?;
+    // Don't offer the assist when cursor is at the end, outside the block itself.
+    let cursor_range = TextRange::from_to(node.range().end(), node.range().end());
```

Comment:
I guessed that `node.range().end()` is used in this context based on the `test_cursor_after_empty_impl_block` test. Is there a good way to get the correct cursor position?

(`ctx.frange` returns a range but IIUC if it's a selection then there's no telling whether the cursor is at the selection start or its end)

Reply:
>Is there a good way to get the correct cursor position?

No, I don't even think it is exposed via LSP API. 

Actually, I don't think we need to special case "cursor is just after the element" case, I think we should just make the intention available. In fact, IntelliJ has special code to make this work:

https://github.com/JetBrains/intellij-community/blob/fe455792c18fc0e2ee81401f3a55710521b503b3/platform/lang-api/src/com/intellij/codeInsight/intention/BaseElementAtCaretIntentionAction.java

I think we probably should even use `let impl_node: ast::ImplBlock = ctx.node_at_offset()` here: it also handles the case if the cursor is just between the two nodes.

Reply:
Done, implemented the `node_at_offset` approach